Mar ipns curriculum (#188)
protocol/launchpad: docs/ipfs/mutable-content.md

- 5. **Sequence** (uint64)
  - Represents the current version of the record (starts at 0)
- 6. **Public Key** (bytes)
  - Public key used to sign this record
  - Note: The public key **must** be included if it cannot be extracted from the peer ID (reference [libp2p/specs#100](https://github.com/libp2p/specs/pull/100/files)).
- 7. **ttl** (uint64)
  - A hint for how long the record should be cached before going back to, for instance the DHT, in order to check if it has been updated.

These records are stored locally, as well as spread across the network, in order to be accessible to everyone. For storing this structured data, we use [Protocol Buffers](https://github.com/google/protobuf), which is a language-neutral, platform neutral extensible mechanism for serializing structured data.

```
message IpnsEntry {
	enum ValidityType {
		// setting an EOL says "this record is valid until..."
		EOL = 0;
	}
	required bytes value = 1;
	required bytes signature = 2;
	optional ValidityType validityType = 3;
	optional bytes validity = 4;
	optional uint64 sequence = 5;
	optional uint64 ttl = 6;
	optional bytes pubKey = 7;
}

> ```

Reviewer: nit:no need to copy this here, just link to specs at  https://github.com/ipfs/specs/blob/main/IPNS.md#ipns-record (we will update this to v2 soon, so you would have outdated details here)

Author: :thumbsup::thumbsup:

### Pinning and Immutability

By default, content on IPFS is not _pinned_. That means when you add a file to IPFS, it will eventually stop being discoverable on the network. To mitigate this problem you must _pin_ your content to your [node](https://docs.ipfs.tech/concepts/nodes/#nodes) to ensure it persistents on IPFS. 

Your options for pinning content are:
1. You pin it locally to your own node. Note that if the content is only pinned to your local node, it must be online for peers to get that content.
2. You use a [pinning service](https://docs.ipfs.io/concepts/persistence/#pinning-services). Some pinning services require payment for their services, others offer a [free storage allowance for new users](https://docs.ipfs.tech/concepts/persistence/#ipfs-filecoin-solutions).

![Pinning IPFS](<../../.gitbook/assets/pinning.png>)

### Mutable Content | ResNetLabs on Tour – David Dias
_See the full set of resources [on the ResNetLab Tutorials page](https://research.protocol.ai/tutorials/resnetlab-on-tour)_

{% embed url="https://www.youtube.com/watch?v=57guoGS53Bo" %}

## The InterPlanetary Name System (IPNS)

> Since IPFS uses CIDs, if you were to share an IPFS address such as `/ipfs/QmbezGequPwcsWo8UL4wDF6a8hYwM1hmbzYv2mnKkEWaUp` with someone, you would need to give the person a new link every time you update the content.

Reviewer 2: ```suggestion
Since IPFS uses CIDs, if you were to share an IPFS address such as `/ipfs/QmbezGequPwcsWo8UL4wDF6a8hYwM1hmbzYv2mnKkEWaUp` with someone, you would need to give the person a new link every time you update the content, because every change would result in a new CID.
```

> The [InterPlanetary Name System (IPNS)](https://docs.ipfs.io/concepts/ipns/#interplanetary-name-system-ipns) solves this issue by creating a link that can be updated. IPNS addresses a very important point, to help bridge the unfamiliarity gap to a new internet. For example, when you go to a users' website, you expect to find that same website with the same link or URL in the future. If that website is updated, you will see those updates as well. This "link" in IPNS is called a _name_, this _name_ is the hash of a public key. The name is associated with a record containing information about the hash that it points to and is signed by the public key’s corresponding private key. This allows new records to be signed and published at any time. Using IPNS means that when someone searches for your website using your _name_, they will receive the most up-to-date content as expected in today's internet. You can learn more about IPNS and [how to use it here](https://docs.ipfs.tech/concepts/ipns/#example-ipns-setup-with-cli). 

Reviewer 2: > IPNS addresses a very important point, to help bridge the unfamiliarity gap to a new internet.

I think this sentence doesn't contribute much here to understanding IPNS. 

Some refinement suggestions

```suggestion
The [InterPlanetary Name System (IPNS)](https://docs.ipfs.io/concepts/ipns/#interplanetary-name-system-ipns) solves this issue by creating a link that can be updated. In the location addressed web, when you go to a user's website, you expect the **link** or URL to remain the same even if that website is updated. In other words, the **link** is a pointer to something that can change. With IPNS, such a mutable link is called a **name** and is the hash of a public key. Every mutable IPNS name is associated with a public/private key pair.

An IPNS name is associated with a record containing the CID it points to and a signature by the public key’s corresponding private key. This allows new records to be signed and published at any time. Using IPNS means that when someone searches for your website using your IPNS _name_, they will receive the most up-to-date content as expected on today's internet. You can learn more about IPNS and [how to use it here](https://docs.ipfs.tech/concepts/ipns/#example-ipns-setup-with-cli). 
```

> When looking up an IPNS _name_, use the `/ipns/` prefix:

Reviewer 2: As far as I know, this mostly applies to path gateway URLs:
```suggestion
When looking up an IPNS _name_ via an IPFS gateway, use the `/ipns/` prefix:
```

```

> /ipns/QmSrPmbaUKA3ZodhzPWZnpFgcPMFWF4QsxXbkWfEptTBJd

Author: ```suggestion
/ipns/k2k4r8jl0yz8qjgqbmc2cdu5hkqek5rj6flgnlkyywynci20j0iuyfuj
```

```

### Deep Dive into IPNS

> IPNS is a self-certifying mutable pointer. Meaning any name that gets published is signed by a private key and anyone else can verify that it was signed by that peer with just the _name_. This self-certifying nature gives IPNS a number of super-powers not present in consensus systems (DNS, blockchain identifiers, etc.), some notable ones include: mutable link information can come from anywhere, not just a particular service/system, and it is very fast and easy to confirm a link is authentic. IPNS introduces a new data structure to IPFS called an _IPNS Record_. This gives you added functionality of controlling expiration of some CID content and version numbering. Checkout the [IPNS spec](https://github.com/ipfs/specs/blob/main/IPNS.md#ipns-record) to learn more about IPNS records.

Reviewer 2: ```suggestion
IPNS is a self-certifying mutable pointer. Meaning any name that gets published is signed by a private key and anyone else can verify that it was signed by that peer with just the _name_. This self-certifying nature gives IPNS a number of super-powers not present in consensus systems (DNS, blockchain identifiers, etc.), some notable ones include: mutable link information can come from anywhere, not just a particular service/system, and it is very fast and easy to confirm a link is authentic. IPNS names are encapsulated by a data structure called an _IPNS Record_ which contains the expiration date of a name and the sequence number which is incremented whenever the record is updated. Check out the [IPNS spec](https://github.com/ipfs/specs/blob/main/IPNS.md#ipns-record) to learn more about IPNS records.
```

#### How it works

> * Publishing - When you first publish something with IPNS. You are creating a brand new record, you are entering a CID to point to, and signing the record to establish you are the owner and certify the results. This gets published to your local datastore, then the DHT.

Reviewer 2: ```suggestion
* Publishing - When you first publish an IPNS name, you create a brand new record containing the CID to point to, expiration timestamp, and a cryptographic signature of the record created with the private key to establish you are the owner and certify the name. This gets published to your local datastore, then the DHT.
```

> * Republishing - On each update to a record, including one that only refreshes its validity, you have to first find the record you are updating, then create a brand new record based off that information. You update the `sequence` number (add 1), update the `value` (with new CID), re-sign the `signature` field, and refill the other options.

Reviewer 2: ```suggestion
* Updating - Whenever you update a record, including to refresh its validity, you create a new record based on the previous, make changes to any of the fields (`CID`, `expiry`), increment the sequence number, add a new signature with the private key and then publish it to the network. 
```

Author: @2color Counter suggestion. Thoughts? 
``` suggestion
Whenever you update a record, including to refresh its validity, you create a new record based on the previous, update the CID pointer if needed, increment the sequence number, then sign and publish it to the network.
```

Reviewer 2: Even better :)

* Searching - [Kubo](https://github.com/ipfs/kubo) uses the DHT to find peers that will have the queried record. This method has gotten faster over the years, but is still limited by the speed of DHT resolution itself. Alternative transports can be used to improve resolution speed (see [PubSub](#pubsub--ipns)).

> * Validity - A record is valid for 24 hours by default, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to DHT to find a new, valid version of your record. Issues with validity don't arise if you're the original publisher of a record:  Kubo will periodically update validity and republish your record to local record store and to DHT to keep it alive for you.

Reviewer 2: ```suggestion
* Validity - A record is valid for 24 hours by default, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to DHT to find a new, valid version of your record. Kubo will periodically update validity and publish the updated IPNS records to the local record store and to DHT to keep it alive. DHT nodes will drop stored values after 24 jours, no matter what the validity. For this reason, for the IPNS name to be resolvable, the record needs to continuously republished
```

Author: ```suggestion
* Validity - A record is valid for 24 hours by default, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to DHT to find a new, valid version of your record. Issues with validity don't arise if you're the original publisher of a record:  Kubo will periodically update validity and publish the updated IPNS records to the local record store and to DHT to keep it alive. DHT nodes will drop stored values after 24 hours, no matter what the validity. For this reason, for the IPNS name to be resolvable, the record needs to be continuously republished.
```

> * Keys - A _name_ is a CID of a public key in a key pair. The public key is used by Kubo for identifying your peer ([PeerID](https://docs.ipfs.tech/concepts/glossary/#peer-id)). You can [generate new sets of keys with Kubo](https://docs.ipfs.tech/reference/kubo/cli/#ipfs-key-gen) and use them to create different IPNS records. 

Reviewer 2: ```suggestion
* Keys - A _name_ is a hash of a public key in a key pair. By default, the first public key used by Kubo is the same as the one for identifying your peer ([PeerID](https://docs.ipfs.tech/concepts/glossary/#peer-id)). You can [generate new key pairs with Kubo](https://docs.ipfs.tech/reference/kubo/cli/#ipfs-key-gen) and use them to create additional IPNS records. 
```

Author: @2color Love the rewording here. Only thing I wouldn't change is that a _name_ IS a CID. The leading character(s) identify what base encoding is being used to convert the hash of the public key to something shorter. 
ex: https://cid.ipfs.tech/#k2k4r8jl0yz8qjgqbmc2cdu5hkqek5rj6flgnlkyywynci20j0iuyfuj

This ipns name used as an exmaple is found on line 33 here. You can see here that it is actually a CIDv1 of a libp2p pubkey :thumbsup:

> #### Common Pitfalls

Reviewer 2: @aschmahmann Can you please confirm these?

* Resolving IPNS over DHT may be slow. This is due to having the `sequence` number in a record. If there are multiple versions of a name, then you want to make sure that you have the latest version. And you do this by performing a time-bound search through the DHT for peers that hold your records and checking a quorum for  the latest one. Kubo will spend up to a minute to try to find at least 16 peers to form a quorum.

> * Users trying to use IPNS from JavaScript in web browsers run into a variety of issues here related to how web browsers are isolated from the p2p network due to transport restrictions. To get around this problem, you would delegate the work out to a local or remote node.

Reviewer 2: Would be good if we could actually link to how this delegation works. For most purposes, I can imagine that users resolving IPNS in a browser would use a gateway.

Reviewer: @2color Yes, it will be the first one in  [`X-Ipfs-Roots` (response header)](https://github.com/ipfs/specs/blob/main/http-gateways/PATH_GATEWAY.md#x-ipfs-roots-response-header).

Important to note here that this is not "Trusted" way of doing this (client trusts gateway did not lie about CID). 
The proper way will be to add new response type to gateway to return a signed IPNS record, that way client will be able to verify IPNS2CID binding without having to trust gateway (we dont have that yet).

Author: ```suggestion
* Users trying to use IPNS from [JavaScript in web browsers](https://github.com/ipfs/js-ipfs/blob/master/docs/BROWSERS.md) run into a variety of issues here related to how web browsers are isolated from the p2p network due to transport restrictions. To get around this problem, you would delegate the work out to a local or remote node. Another workaround would be using [public gateways](https://docs.ipfs.tech/concepts/ipfs-gateway/#public-gateways) for resolving [IPNS records and IPFS CIDs](https://protocol-labs.gitbook.io/launchpad-curriculum/launchpad-learning-resources/ipfs/ipfs-gateways#path-gateway).
```

> * Currently, there is no way to revalidate someone elses record to keep it from expiring. You **are** able to re-provide someone elses record, but it will only be valid for as long as the original publisher specified in the `lifetime` field of their record. 

Reviewer 2: I would rephrase this to use consistent nomenclature as above (avoiding revalidating):

```suggestion
* Only the holder of the corresponding private key can update an IPNS record, e.g. to extend its validity. On the other hand, anyone is able to re-provide someone other's IPNS records as long as the timestamp set in the `validity` field by the original publisher is in the future.
```

Author: ```suggestion
* Only the holder of the corresponding private key can update an IPNS record, e.g. to extend its validity. Regardless, any peer can act as a host and provide the record to the network if that record is valid.
```

* Safe Vs. Old problem. This is a question around practicality on IPNS content. If I accidentally let my record expire, do I want my users to fetch and resolve the data anyways (aka serve **old** data) or not at all (**safe**). Currently, the default is to be safe, and not serve any data. This is an ongoing and situational conversation, you can [read about it on GitHub](https://github.com/ipfs/kubo/issues/1958#issuecomment-444201606)

Sources: [Overview of IPNS by Adin](https://pl-strflt.notion.site/IPNS-Overview-and-FAQ-071b9b14f12045ea842a7d51cfb47dff#0963fe6b470a4c55b1929146c360dc95), [IPNS Spec](https://github.com/ipfs/specs/blob/main/IPNS.md), [Discuss forum for IPFS](https://discuss.ipfs.tech/t/how-do-i-make-my-ipns-records-live-longer/14768/17?u=lidel)

### DNSLink

> [DNSLink](https://docs.ipfs.tech/concepts/glossary/#dnslink) leverages the powerful distributed architecture of DNS for a variety of systems that require internet scale mutable names or pointers. With DNSLink, you can store a link to content-addressed data using any DNS domain name. This is currently much faster than IPNS and also uses human-readable names. It can point at any content path (`/ipfs/<cid>` or `/ipns/<libp2p-key>`).  By using IPNS name in a DNSLink record, one can keep DNS intact and publish updates over IPNS. First, put the IPNS name in a TXT record at a specific subdomain. Then, you can resolve the name from any program by looking up the TXT value. Your programs and systems can parse out the record value, and follow the CID wherever it may go. 

Reviewer 2: ```suggestion
DNSLink is a standard for a mutable pointer system that stores its links in DNS TXT 
records corresponding to a given domain name. With DNSLink you can use a mutable pointer like a DNS domain name to point to content-addressed data in IPFS.

The benefits of DNSLink include:
- Human-readable mutable pointers, e.g. `/ipns/libp2p.io`.
- DNSLink is that it leverages the distributed architecture of DNS to enable internet-scale mutable names or pointers which interoperate with IPFS. 
- In many cases, resolving a DNSLink pointer is faster than IPNS. 

To create a mutable pointer with DNSLink, you need:
- a DNS domain name you control
- A CID or an IPNS name to link to

For example, `libp2p.io` has a TXT DNS record `_dnslink.libp2p.io` with the value `dnslink=/ipfs/bafybeibwlrl7olq5sggibzucp5s6y5n22numfpyjlzntnuvgs5zt2umjuu`. This DNS record can be _mutated_ at any point to point to a new CID.

Finally, there are several common ways to resolve DNSLink names:
- With an IPFS gateway like an IPNS name, e.g. https://ipfs.io/ipns/libp2p.io or https://cloudflare-ipfs.com/ipns/libp2p.io
- Directly via the domain, but this requires some additional DNS configuration as explained in the [DNSLink docs](https://dnslink.dev/#example-ipfs-gateway).
 
> Note that while using DNSLink to link to an IPNS name is possible, it's often counterintuitive, because you are using a mutable pointer (DNS record) to point to another mutable pointer (IPNS name). 
```

**You can follow tutorials and read more at [https://dnslink.dev](https://dnslink.dev/)**

#### Quick explanation of dnslink in IPFS

{% embed url="https://www.youtube.com/watch?v=YxKZFeDvcBs" %}

Source: [Introduction to DNSlink](https://dnslink.dev/#introduction), [IPFS docs on DNSLink](https://docs.ipfs.tech/concepts/dnslink/)

### Pubsub + IPNS

> [**Publish/Subsribe (PubSub)**](https://docs.libp2p.io/concepts/publish-subscribe/) is a messaging protocol to quickly communicate with other peers. Whenever one Publishes some content, Subscribing peers will receive it almost instantly. This protocol is not specific to IPFS or IPNS, but to [Libp2p](https://docs.ipfs.tech/concepts/libp2p/); to allow for quick delivery of content over a network of users. For this reason, the two teams got together to bring instant messaging to a system where sharing and receiving content takes too long. With PubSub enabled on IPNS, updates to a record can be shared virtually instantly with subscribers. 

Reviewer 2: To be precise here: 
- The name of the protocol is gossipsub
- It's a general-purpose pubsub protocol. 
- It allows for the quick delivery of **messages** (message being the word used by the spec rather than _content_) without flooding/overloading the network (Spec uses "moderate amplification factors")

> For this reason, the two teams got together to bring instant messaging to a system where sharing and receiving content takes too long.

I don't think this adds much in this context.

Author: ```suggestion
[**Publish/Subsribe (PubSub)**](https://docs.libp2p.io/concepts/publish-subscribe/) is a messaging protocol to quickly communicate with other peers. Whenever a peer Publishes a message, Subscribing peers will receive it almost instantly. This protocol is not specific to IPFS or IPNS, but to [Libp2p](https://docs.ipfs.tech/concepts/libp2p/); paired with IPNS, it allows for quick delivery of records over the network. With PubSub enabled on IPNS, updates to a record can be shared virtually instantly with subscribers. 
```

```
/ipns/QmSrPmbaUKA3ZodhzPWZnpFgcPMFWF4QsxXbkWfEptTBJd
```

### Deep Dive into IPNS

IPNS is a self-certifying mutable pointer. Meaning any name that gets published is signed by a private key and anyone else can verify that it was signed by that peer with just the _name_. This self-certifying nature gives IPNS a number of super-powers not present in consensus systems (DNS, blockchain identifiers, etc.), some notable ones include: mutable link information can come from anywhere, not just a particular service/system, and it is very fast and easy to confirm a link is authentic. IPNS introduces a new data structure to IPFS called a _Record_. This gives you added functionality of controlling expiration of some CID content and version numbering. Checkout the [IPNS spec](https://github.com/ipfs/specs/blob/main/IPNS.md#ipns-record) to learn more about records.

#### How it works

* Publishing - When you first publish something with IPNS. You are creating a brand new record, you are entering a CID to point to, and signing the record to establish you are the owner and certify the results. This gets publish to your local datastore, then the DHT.

> * Republishing - On each update to a record, you have to first find the record you are updating, then create a brand new record based off that information. You update the `sequence` number (add 1), update the `value` (with new CID), re-sign the `signature` field, and refill the other options.

Reviewer: ```suggestion
* Republishing - On each update to a record, including one that only refreshes its validity, you have to first find the record you are updating, then create a brand new record based off that information. You update the `sequence` number (add 1), update the `value` (with new CID), re-sign the `signature` field, and refill the other options.
```

> * Searching - Kubo uses the DHT to find peers that will have the queried record. This method has gotten faster over the years, but is still generally slow to reslove a name.

Reviewer: ```suggestion
* Searching - [Kubo](https://github.com/ipfs/kubo) uses the DHT to find peers that will have the queried record. This method has gotten faster over the years, but is still limited by the speed of DHT resolution itself. Alternative transports can be used to improve resolution speed (see [PubSub](#pubsub--ipns)).
```

> * Validity - A record is valid for 24 hours by default, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to their DHT to find your address and grab the latest valid version of your record. Issues with validity don't arise if you're the original publisher of a record. Kubo will automatically republish your record on your local node to keep it alive if you are online.

Reviewer: ```suggestion
* Validity - A record is valid for 24 hours by default, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to DHT to find a new, valid version of your record. Issues with validity don't arise if you're the original publisher of a record:  Kubo will periodically update validity and republish your record to local record store and to DHT to keep it alive for you.
```

> * Keys - A _name_ is the hash of a public key in a key pair. The public key is used by Kubo for identifying your peer (peerID). You can generate new sets of keys with Kubo and use them create different IPNS records. 

Reviewer: ```suggestion
* Keys - A _name_ is a CID of a public key in a key pair. The public key is used by Kubo for identifying your peer ([PeerID](https://docs.ipfs.tech/concepts/glossary/#peer-id)). You can [generate new sets of keys with Kubo](https://docs.ipfs.tech/reference/kubo/cli/#ipfs-key-gen) and use them to create different IPNS records. 
```

#### Common Pitfalls

> * IPNS on DHT is slow. This is due to having the `sequence` number in a record. If there are multiple versions of a name, then you want to make sure that you have the latest version. And you do this by searching through the DHT for peers that hold your record and checking the version number for the latest one.

Reviewer: ```suggestion
* Resolving IPNS over DHT may be slow. This is due to having the `sequence` number in a record. If there are multiple versions of a name, then you want to make sure that you have the latest version. And you do this by performing a time-bound search through the DHT for peers that hold your records and checking a quorum for  the latest one. Kubo will spend up to a minute to try to find at least 16 peers to form a quorum.
```

> * Another issue with the DHT is that after so much content has been added to it, it will take longer and longer to resolve your record over time. Assuming the garbage collector hasn't already removed it. It is not uncommon for IPNS to return you nothing when fetching a record.

Reviewer: I am confused by this, sounds like a bunch of unrelated things clustered together without enough detail to be useful to the reader. 
 Either elaborate on each point, or remove this entirely.

```suggestion
```

* Users trying to use IPNS from JavaScript in web browsers run into a variety of issues here related to how web browsers are isolated from the p2p network due to transport restrictions. To get around this problem, you would delegate the work out to a local or remote node.
* Currently, there is no way to revalidate someone elses record to keep it from expiring. You **are** able to re-provide someone elses record, but it will only be valid for as long as the original publisher specified in the `lifetime` field of their record. 

> * When users force third party republishing by sharing private keys, the issue of ["sequence number collision"](https://discuss.ipfs.tech/t/ipns-beyond-the-basics-no-ipns-pinning-service-any-docs-on-this/13424/2) would occur (among various others). This would be when two people updating a shared record have the same `sequence` number, but different CIDs. How would a receiver of said record know which one is correct?

Reviewer: No such thing as a third party, every party becomes the first party if they have private key :)

```suggestion
* When users share private keys and publish from multiple nodes, the issue of ["sequence number collision"](https://discuss.ipfs.tech/t/ipns-beyond-the-basics-no-ipns-pinning-service-any-docs-on-this/13424/2) would occur (among various others). This would be when two people updating a shared record have the same `sequence` number, but different CIDs. How would a receiver of said record know which one is correct?
```

* Safe Vs. Old problem. This is a question around practicality on IPNS content. If I accidentally let my record expire, do I want my users to fetch and resolve the data anyways (aka serve **old** data) or not at all (**safe**). Currently, the default is to be safe, and not serve any data. This is an ongoing and situational conversation, you can [read about it on GitHub](https://github.com/ipfs/kubo/issues/1958#issuecomment-444201606)

Sources: [Overview of IPNS by Adin](https://pl-strflt.notion.site/IPNS-Overview-and-FAQ-071b9b14f12045ea842a7d51cfb47dff#0963fe6b470a4c55b1929146c360dc95), [IPNS Spec](https://github.com/ipfs/specs/blob/main/IPNS.md), [Discuss forum for IPFS](https://discuss.ipfs.tech/t/how-do-i-make-my-ipns-records-live-longer/14768/17?u=lidel)

### DNSLink

> DNSLink leverages the powerful distributed architecture of DNS for a variety of systems that require internet scale mutable names or pointers. With DNSLink, you can store an IPNS link using any DNS domain name. This is currently much faster than IPNS and also uses human-readable names. First, put the IPNS name in a TXT record at a specific subdomain. Then, you can resolve the name from any program by looking up the TXT value. Your programs and systems can parse out the record value, and follow the CID wherever it may go. 

Reviewer: ```suggestion
[DNSLink](https://docs.ipfs.tech/concepts/glossary/#dnslink) leverages the powerful distributed architecture of DNS for a variety of systems that require internet scale mutable names or pointers. With DNSLink, you can store a link to content-addressed data using any DNS domain name. This is currently much faster than IPNS and also uses human-readable names. It can point at any content path (`/ipfs/<cid>` or `/ipns/<libp2p-key>`).  By using IPNS name in a DNSLink record, one can keep DNS intact and publish updates over IPNS. First, put the IPNS name in a TXT record at a specific subdomain. Then, you can resolve the name from any program by looking up the TXT value. Your programs and systems can parse out the record value, and follow the CID wherever it may go. 
```

**You can follow tutorials and read more at [https://dnslink.dev](https://dnslink.dev/)**

#### Quick explanation of dnslink in IPFS

{% embed url="https://www.youtube.com/watch?v=YxKZFeDvcBs" %}

Source: [Introduction to DNSlink](https://dnslink.dev/#introduction), [IPFS docs on DNSLink](https://docs.ipfs.tech/concepts/dnslink/)

### Pubsub + IPNS

[**Publish/Subsribe (PubSub)**](https://docs.libp2p.io/concepts/publish-subscribe/) is a messaging protocol to quickly communicate with other peers. Whenever one Publishes some content, Subscribing peers will receive it almost instantly. This protocol is not specific to IPFS or IPNS, but to [Libp2p](https://docs.ipfs.tech/concepts/libp2p/); to allow for quick delivery of content over a network of users. For this reason, the two teams got together to bring instant messaging to a system where sharing and receiving content takes too long. With PubSub enabled on IPNS, updates to a record can be shared virtually instantly with subscribers. 

> To accomplish IPNS over PubSub, a persistence layer had to be added to the messaging protocol. This was started with the introduction of the [fetch protocol](https://github.com/libp2p/specs/tree/master/fetch). Now when you ask for a name, you are subcribing to that name (called a "topic" in libp2p pubsub), you create a connection with that publisher, then they send you the latest version of the record. The key differentiating factor between IPNS-over-PubSub and IPNS-over-the-DHT (the default behavior) is opening a streaming connection between peers. This way, peers are sending the latest record directly from their local node, as opposed to the default behavior of searching the DHT for the peer with latest version of a record. This means records shared over Pubsub are not available on the DHT and vise versa, unless the publisher opts-in to publish records to both routing options. If you would like to activate IPNS over Pubsub on your Kubo node, you can check out the [UsePubsub option](https://github.com/ipfs/kubo/blob/master/docs/config.md#ipns) in the IPFS config file.

Reviewer: Small nits:
-  `fetch` is implementation detail, more important, is persistence scheme described in specs (added link)
-  does not have to be the original publisher, any peer interested in name-based topic will do

```suggestion
To accomplish IPNS over PubSub, a [persistence layer had to be added](https://github.com/ipfs/specs/blob/main/naming/pubsub.md#layering-persistence-onto-libp2p-pubsub). Now when you ask for a name, you are subcribing to a PubSub topic based on that name, you create a connection with a peer that is following the same name, then they send you the latest version of the record. The key differentiating factor between IPNS-over-PubSub and IPNS-over-the-DHT (the default behavior) is opening a streaming connection between peers. This way, peers are sending the latest record directly from their local node, as opposed to the default behavior of searching the DHT for the peer with latest version of a record. This means records shared over Pubsub are not available on the DHT and vise versa, unless the publisher opts-in to publish records to both routing options. If you would like to activate IPNS over Pubsub on your Kubo node, you can check out the [`Ipns.UsePubsub` option](https://github.com/ipfs/kubo/blob/master/docs/config.md#ipns) in the config file.
```

Since IPFS uses CIDs, if you were to share an IPFS address such as `/ipfs/QmbezGequPwcsWo8UL4wDF6a8hYwM1hmbzYv2mnKkEWaUp` with someone, you would need to give the person a new link every time you update the content.

The [InterPlanetary Name System (IPNS)](https://docs.ipfs.io/concepts/ipns/#interplanetary-name-system-ipns) solves this issue by creating an address that can be updated. IPNS addresses a very important point; to help bridge the unfamiliarity gap to a new internet. For example, when you go to a users' profile on a social media site today, you expect to find it with the same link or URL in the future. And if there are any updates, then you expect to see those updates as well. IPNS builds on top of IPFS, by giving you a "username"; when someone asks for your username, they'll receive the most up-to-date content you published on IPFS. If you change your username, you are generating a brand new CID and by extension you are creating a brand new online profile.

> In IPNS, these user identifiers are not called usernames, rather, they are called _name_'s. A _name_ is the [hash](https://docs.ipfs.io/concepts/hashing/) of a public key. It is associated with a _record_ containing information about the CID it links to that is signed by the corresponding private key. New records can be signed and published at any time.

Reviewer 2: ```suggestion
In IPNS, these user identifiers are not called usernames, rather, they are called _name_'s. A _name_ is the [hash](https://docs.ipfs.io/concepts/hashing/) of a public key. It is associated with a _record_ containing information about the CID it links to that is signed by the corresponding private key. New records can be signed and published at any time by the holder of the private key.
```

When looking up an IPNS address, use the `/ipns/` prefix:

```
/ipns/QmSrPmbaUKA3ZodhzPWZnpFgcPMFWF4QsxXbkWfEptTBJd
```

### Deep: Into IPNS

> IPNS is a self-certifying mutable pointer. Meaning any name that gets published is signed by a private key and anyone else can verify that it was signed by that peer with just the _name_. This self-certifying nature gives IPNS a number of super-powers not present in consensus systems (DNS, blockchain identifiers, etc.), some notable ones include: mutable link information can come from anywhere, not just a particular service/system, and it is to confirm a link is authentic. IPNS introduces a new data structure to IPFS called a _Record_. This gives you added functionality of controlling expiration of some CID content and version numbering. This is what an IPNS object contains:

Reviewer 2: > mutable link information can come from anywhere, not just a particular service/system, and it is to confirm a link is authentic

Not sure I understand the last part of this sentence

```
message IpnsEntry {
	enum ValidityType {
		// setting an EOL says "this record is valid until..."
		EOL = 0;
	}
	required bytes value = 1;
	required bytes signature = 2;
	optional ValidityType validityType = 3;
	optional bytes validity = 4;
	optional uint64 sequence = 5;
	optional uint64 ttl = 6;
	optional bytes pubKey = 7;
}
```

#### How it works

* Publishing - When you first publish something with IPNS. You are creating a brand new record, you are entering a CID to point to, and signing the record to establish you are the owner and certify the results. This gets publish to your local datastore, then the DHT.
* Republishing - On each update to a record, you have to first find the record you are updating, then create a brand new record based off that information. You update the `sequence` number (add 1), update the `value` (with new CID), re-sign the `signature` field, and refill the other options.
* Searching - Kubo uses the DHT to find peers that will have the queried record. This method has gotten faster over the years, but is still generally slow to reslove a name.

> * Validity - A record is only valid for 24 hrs, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to their DHT to find your address and grab the latest valid version of your record. Issues with validity don't arise if you're the original publisher of a record. Kubo will automatically republish your record on your local node to keep it alive if you are online.

Reviewer 2: ```suggestion
* Validity - A record is valid for 24 hrs by default, but you can change its `validity` to be longer. When someone has your record, but it is expired, they will have to go to their DHT to find your address and grab the latest valid version of your record. Issues with validity don't arise if you're the original publisher of a record. Kubo will automatically republish your record on your local node to keep it alive if you are online.
```

> IPNS is a self-certifying mutable pointer. Meaning any name that gets published is signed by a private key and anyone else can verify that it was signed by that peer with just the _name_. This self-certifying nature gives IPNS a number of super-powers not present in consensus systems (DNS, blockchain identifiers, etc.), some notable ones include: mutable link information can come from anywhere, not just a particular service/system, and it is to confirm a link is authentic. IPNS introduces a new data structure to IPFS called a _Record_. This gives you added functionality of controlling expiration of some CID content and version numbering. This is what an IPNS object contains:

Author: Source: https://pl-strflt.notion.site/IPNS-Overview-and-FAQ-071b9b14f12045ea842a7d51cfb47dff - last sentence of first paragraph.

```suggestion
IPNS is a self-certifying mutable pointer. Meaning any name that gets published is signed by a private key and anyone else can verify that it was signed by that peer with just the _name_. This self-certifying nature gives IPNS a number of super-powers not present in consensus systems (DNS, blockchain identifiers, etc.), some notable ones include: mutable link information can come from anywhere, not just a particular service/system, and it is very fast and easy to confirm a link is authentic. IPNS introduces a new data structure to IPFS called a _Record_. This gives you added functionality of controlling expiration of some CID content and version numbering. This is what an IPNS object contains:
```

> * Keys - A _name_ is the hash of a public key pair. This key pair is the libp2p-key used in Kubo for identifying a peer (peerID). So in actuality, when you're searching for a name, you are searching for a shortened and case-insensitive version of a specific peerID. This is possible because names are in [base 36 (k)](https://github.com/multiformats/multibase/blob/master/rfcs/Base36.md) and this base encoding enables support for querying content through browser gateways like [ipfs.io/ipns/k51...](https://ipfs.io)

Reviewer 2: ```suggestion
* Keys - A _name_ is the hash of a public key of a key pair. The first key pair is the key used by Kubo for identifying your peer (peerID). You can create multiple IPNS records by generating additional keys.
```

> So in actuality, when you're searching for a name, you are searching for a shortened and case-insensitive version of a specific peerID.

This is true for the first name. I would add a note that you can create additional names by generating new key pairs.

> So in actuality, when you're searching for a name, you are searching for a shortened and case-insensitive version of a specific peerID.

This is only true for the first name.

> This is possible because names are in [base 36 (k)](https://github.com/multiformats/multibase/blob/master/rfcs/Base36.md) and this base encoding enables support for querying content through browser gateways like [ipfs.io/ipns/k51...](https://ipfs.io)

This feels unnecessary here.

Author: Counter suggestion
```suggestion
* Keys - A _name_ is the hash of a public key in a key pair. The public key is used by Kubo for identifying your peer (peerID). You can generate new sets of keys with Kubo and use them create different IPNS records. 
```

#### Common Pitfalls
* IPNS on DHT is slow. This is due to having the `sequence` number in a record. If there are multiple versions of a name, then you want to make sure that you have the latest version. And you do this by searching through the DHT for peers that hold your record and checking the version number for the latest one.
* Another issue with the DHT is that after so much content has been added to it, it will take longer and longer to resolve your record over time. Assuming the garbage collector hasn't already removed it. It is not uncommon for IPNS to return you nothing when fetching a record.
* Users trying to use IPNS from JavaScript in web browsers run into a variety of issues here related to how web browsers are isolated from the p2p network due to transport restrictions. To get around this problem, you would delegate the work out to a local or remote node.

> * Currently, there is no way to keep someone elses record alive/valid. This is unfortunate because you may want to keep the original record alive on your node. That way others can also come to you for a valid record and thus expand longterm record discovery.

Reviewer 2: I'm not sure this is correct.

The `signature` is a concatenation of the `value`, `validity` field and `validityType`. As long as the record hasn't expired, i.e. validty is in the future, I'm pretty sure other peers can continue publishing the record. 

https://github.com/ipfs/specs/blob/main/IPNS.md

Author: Ah yes thank you for pointing this out. Also see comment regarding line 93.

> * If third party republishing was possible, the issue of "sequence number collision" would occur. This would be when two people updating a record have the same `sequence` number, but different CIDs. How would anyone know which one is correct?

Reviewer 2: Only the holder of the private key can sign a record, that doesn't necessarily prevent republishing.

Reviewer 2: This is pretty much confirmed in this thread https://github.com/ipfs/kubo/issues/1958#issuecomment-1026322601

Reviewer 2: Ok this is definitely confirmed [here](https://github.com/ipfs/specs/blob/main/naming/pubsub.md)

> An inherent property of the IPNS PubSub Router is that IPNS records are republishable by peers other than the peer that originated the record. This implies that as long as a peer on the network has an IPNS record it can be made available to other peers (although the records may be ignored if they are received after the IPNS record's End-of-Life/EOL).

Author: aahhh okay I hadn't seen that before. I was basing this line93 off this forum question/response: https://discuss.ipfs.tech/t/ipns-beyond-the-basics-no-ipns-pinning-service-any-docs-on-this/13424/2

> However this nonce counter is local to a node.
If you have two nodes publishing under the same key, the first one can publish using nonce 100 and the second one, unaware that the first one already used 100, will also use nonce 100.
Giving you this situation:

I was trying to describe a scenario, that if two nodes have access to the same private key, they can both sign the same record. and this can cause problems.

But since, a third party can republish a record without revalidating it/resigning, then I do need to change up line 92&93 :thumbsup:

* To authorize third party republishing, you would share your private key with someone else. They would `ipfs key import` it into their node and could start keeping your record alive aswell. But now they also have full permissions to impersonate you and act maliciously.
* Safe Vs. Old problem. This is a question around practicality on IPNS content. If I accidentally let my record expire, do I want my users to fetch and resolve the data anyways (aka serve **old** data) or not at all (**safe**). Currently, the default is to be safe, and not serve any data. This is an ongoing and situational conversation you can [read up on it on github](https://github.com/ipfs/kubo/issues/1958#issuecomment-444201606)

Sources: [Overview of IPNS by Adin](https://pl-strflt.notion.site/IPNS-Overview-and-FAQ-071b9b14f12045ea842a7d51cfb47dff#0963fe6b470a4c55b1929146c360dc95), [IPNS Spec](https://github.com/ipfs/specs/blob/main/IPNS.md)

> ### Pubsub + IPNS

Reviewer 2: I would use the introduction paragraph to already introduce Pubsub and its role with IPNS. 

Pubsub is one of the ways to propagate IPNS records across the IPFS network. I would contrast it to the DHT which follows pull semantics, i.e., you query from the DHT to resolve an IPNS record. I'm not 100% up to date on how IPNS works with pubsub (a topic I'm actively learning), but I think this section would benefit from more details.

Author: :thumbsup: Agreed. Good point  :thumbsup:

> IPNS [solves the issue of creating constant links to content](https://blog.fleek.co/posts/immutable-ipfs) that may be updated, and provides the 'new' link to content that you want to share with others. IPNS uses public-link cryptography to assign a link to the content you created on IPFS.

Reviewer 2: ```suggestion
IPNS [solves the issue of creating constant links to content](https://blog.fleek.co/posts/immutable-ipfs) that may be updated, and provides the 'new' link to content that you want to share with others. IPNS uses public-key cryptography to create a mutable link to immutable content on IPFS.
```
